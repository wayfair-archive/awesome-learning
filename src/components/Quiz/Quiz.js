import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {Button, Typography, Box} from '@material-ui/core';
import {makeStyles, useTheme} from '@material-ui/core/styles';
import {Link} from 'gatsby';
import {
  formatQuestionId,
  getChoiceIndex,
  getQuestionIndex,
} from './quizUtilities';
import Question from './Question';
import questionValidator from './Question/questionValidator';

const useStyles = makeStyles((theme) => ({
  sectionTitle: {
    paddingTop: theme.spacing(3),
  },
  linkTitle: {
    textTransform: 'uppercase',
    textDecoration: 'underline',
    '&:hover, &:focus': {
      textDecoration: 'none',
    },
  },
  paragraphContainer: {
    paddingBottom: theme.spacing(4),
  },
}));

const Quiz = ({quiz, slug, title}) => {
  const classes = useStyles();
  const theme = useTheme();
  const initialState = {
    shouldShowCorrectChoice: false,
    questions: quiz.questions,
  };
  // Hooks to maintain state within this Quiz
  const [{shouldShowCorrectChoice, questions}, updateQuizState] = useState(
    initialState
  );

  // Handles changing of a question; update state based on which choice(s) the user has selected
  const handleInputChange = ({currentTarget: fieldset, target}) => {
    // Get the current question and mark the appropriate choices as being selected by the user
    const questionIndex = getQuestionIndex(target);
    const currentQuestion = questions[questionIndex];
    switch (currentQuestion.type) {
      case 'checkbox': {
        // Gets the indices of selected checkboxes
        const selectedIndices = Array.from(
          fieldset.querySelectorAll('input:checked')
        ).map(getChoiceIndex);

        // Updates every choice based on whether or not it is in the list of user-selected choices
        currentQuestion.choices.forEach((choice, index) => {
          choice.isSelected = selectedIndices.includes(index);
        });
        break;
      }
      case 'radio': {
        // Set the selection status of all question choices based on
        // whether or not the user has selected them
        const choiceIndex = getChoiceIndex(target);
        currentQuestion.choices.forEach((choice, index) => {
          choice.isSelected = choiceIndex === index;
        });
        break;
      }
      default: {
        break;
      }
    }
    // Clone the questions array to force a state change
    const updatedQuestions = Array.from(questions);
    // Update the state of the quiz
    updateQuizState({
      shouldShowCorrectChoice: false,
      questions: updatedQuestions,
    });
  };

  // Handles the submission of the quiz
  const handleQuizSubmit = () => {
    // Scroll to the top of the page
    window.scrollTo(0, 0);
    // Validate every question and update the state of the quiz
    const validatedQuestions = questions.map(questionValidator);
    updateQuizState({
      shouldShowCorrectChoice: true,
      questions: validatedQuestions,
    });
  };

  return (
    <Box m="auto" maxWidth={theme.breakpoints.values.lg}>
      <Typography
        variant="h4"
        color="secondary"
        component={Link}
        to={slug}
        className={classes.linkTitle}
      >
        Back to Lesson
      </Typography>
      <Typography variant="h1" className={classes.sectionTitle}>
        {title} Pre-Read Quiz
      </Typography>
      <Box>
        {quiz.description && (
          <Typography variant="body1" className={classes.sectionTitle}>
            {quiz.description}
          </Typography>
        )}
        {/* Iterate over all questions in this Quiz */}
        {questions.map((question, questionIndex) => {
          const questionId = formatQuestionId(questionIndex);
          const questionTitle = `Q${questionIndex + 1}: ${
            question.description
          }`;
          return (
            <Box my={4} key={questionId}>
              <Question
                handleInputChange={handleInputChange}
                question={question}
                questionId={questionId}
                questionIndex={questionIndex}
                shouldShowCorrectChoice={shouldShowCorrectChoice}
                title={questionTitle}
              />
            </Box>
          );
        })}
        <Button
          variant="contained"
          color="primary"
          size="large"
          onClick={handleQuizSubmit}
        >
          Submit Quiz
        </Button>
      </Box>
    </Box>
  );
};

Quiz.propTypes = {
  quiz: PropTypes.shape({
    questions: PropTypes.array,
    description: PropTypes.node,
  }).isRequired,
  slug: PropTypes.string,
  title: PropTypes.string,
};

export default Quiz;
