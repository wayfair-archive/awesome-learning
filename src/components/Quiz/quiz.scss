.Quiz {
  $rootSelector: &;
  display: block;
  margin: auto;
  max-width: $layout-width;

  &-homeButton {
    @include margin(1, 1.3, 0, 1.3);
  }

  &-submitButton {
    border: 1px solid $color--border;
    border-radius: $button-border-radius;
    color: $color--base;
    font-size: $font--1500;
    font-weight: $font-weight--bold;
    height: $button-height;
    display: inline-block;
    line-height: $button-height;
    padding: 0 24px;
    text-align: center;
    @include margin-top(1);
  }
}
