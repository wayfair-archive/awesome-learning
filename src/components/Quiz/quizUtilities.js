// Formats the input ID for one question choice
export const formatChoiceId = (questionId, choiceIndex) =>
  `${questionId}_choice${choiceIndex}`;

// Formats the question ID for one question
export const formatQuestionId = (questionIndex) => `question${questionIndex}`;

// Gets the numeric index of this choice
export const getChoiceIndex = (element) =>
  parseInt(element.getAttribute('data-choice-index'), 10);

// Gets the numeric index of this choice
export const getQuestionIndex = (element) =>
  parseInt(element.getAttribute('data-question-index'), 10);
