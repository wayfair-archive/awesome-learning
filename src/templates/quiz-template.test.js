import React from 'react';
import renderer from 'react-test-renderer';
import Quiz from './quiz-template';

describe('CourseTemplate', () => {
  const props = {
    data: {
      site: {
        siteMetadata: {
          title: 'test',
          subtitle: 'test'
        }
      },
      markdownRemark: {
        html: '<p>test</p>',
        fields: {
          tagSlugs: [
            '/test_0',
            '/test_1'
          ]
        },
        frontmatter: {
          title: 'test',
          preReadQuiz: {
            description: 'test',
            questions: [
              {
                description: 'test1',
                choices: ["testChoice1", "testChoice2", "testChoice3", "testChoice4"],
                correctChoices: [2],
                explanation: 'test',
                type: 'radio'
              },
              {
                description: 'test2',
                choices: ["testChoice5", "testChoice6", "testChoice7", "testChoice8"],
                correctChoices: [0],
                explanation: 'test2',
                type: 'checkbox'
              }
            ]
          }
        }
      }
    }
  };

  it('renders correctly', () => {
    const tree = renderer.create(<Quiz {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
