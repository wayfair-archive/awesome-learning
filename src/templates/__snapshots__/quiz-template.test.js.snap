// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseTemplate renders correctly 1`] = `
Array [
  <StaticQuery
    render={[Function]}
  />,
  <div
    className="Layout"
  >
    <div
      className="Page"
    >
      <div
        className="Page-inner"
      >
        <div
          className="Page-body"
        >
          <div
            className="Quiz"
          >
            <div
              className="Quiz-homeButton"
            >
              <Link
                activeClassName=""
                className="Link Link--tertiary"
              >
                Back to lesson
              </Link>
            </div>
            <div
              className="ContentSection"
            >
              <header
                className="ContentSection-title"
              >
                <h1>
                  test
                </h1>
                <h2>
                  Pre-Read Quiz
                </h2>
              </header>
              <div
                className="ContentSection-content  ContentSection-content--alignLeft"
              >
                <p>
                  test
                </p>
                <fieldset
                  className="Question"
                  onChange={[Function]}
                >
                  <legend
                    className="Question-legend"
                  >
                    Q1: test1
                  </legend>
                  <div
                    className="Question-choice"
                  >
                    <input
                      className="Question-choiceInput"
                      data-choice-index={0}
                      data-question-index={0}
                      id="question0_choice0"
                      name="question0"
                      type="radio"
                    />
                    <label
                      className="Question-choiceLabel"
                      htmlFor="question0_choice0"
                    />
                  </div>
                  <div
                    className="Question-choice"
                  >
                    <input
                      className="Question-choiceInput"
                      data-choice-index={1}
                      data-question-index={0}
                      id="question0_choice1"
                      name="question0"
                      type="radio"
                    />
                    <label
                      className="Question-choiceLabel"
                      htmlFor="question0_choice1"
                    />
                  </div>
                  <div
                    className="Question-choice"
                  >
                    <input
                      className="Question-choiceInput"
                      data-choice-index={2}
                      data-question-index={0}
                      id="question0_choice2"
                      name="question0"
                      type="radio"
                    />
                    <label
                      className="Question-choiceLabel"
                      htmlFor="question0_choice2"
                    />
                  </div>
                  <div
                    className="Question-choice"
                  >
                    <input
                      className="Question-choiceInput"
                      data-choice-index={3}
                      data-question-index={0}
                      id="question0_choice3"
                      name="question0"
                      type="radio"
                    />
                    <label
                      className="Question-choiceLabel"
                      htmlFor="question0_choice3"
                    />
                  </div>
                </fieldset>
                <fieldset
                  className="Question"
                  onChange={[Function]}
                >
                  <legend
                    className="Question-legend"
                  >
                    Q2: test2
                  </legend>
                  <div
                    className="Question-choice"
                  >
                    <input
                      className="Question-choiceInput"
                      data-choice-index={0}
                      data-question-index={1}
                      id="question1_choice0"
                      name="question1"
                      type="checkbox"
                    />
                    <label
                      className="Question-choiceLabel"
                      htmlFor="question1_choice0"
                    />
                  </div>
                  <div
                    className="Question-choice"
                  >
                    <input
                      className="Question-choiceInput"
                      data-choice-index={1}
                      data-question-index={1}
                      id="question1_choice1"
                      name="question1"
                      type="checkbox"
                    />
                    <label
                      className="Question-choiceLabel"
                      htmlFor="question1_choice1"
                    />
                  </div>
                  <div
                    className="Question-choice"
                  >
                    <input
                      className="Question-choiceInput"
                      data-choice-index={2}
                      data-question-index={1}
                      id="question1_choice2"
                      name="question1"
                      type="checkbox"
                    />
                    <label
                      className="Question-choiceLabel"
                      htmlFor="question1_choice2"
                    />
                  </div>
                  <div
                    className="Question-choice"
                  >
                    <input
                      className="Question-choiceInput"
                      data-choice-index={3}
                      data-question-index={1}
                      id="question1_choice3"
                      name="question1"
                      type="checkbox"
                    />
                    <label
                      className="Question-choiceLabel"
                      htmlFor="question1_choice3"
                    />
                  </div>
                </fieldset>
                <button
                  className="Quiz-submitButton"
                  onClick={[Function]}
                >
                  Submit Quiz
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
